src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -648,8 +648,23 @@ public void SendAsync(MailMessage message, object? userToken)
                         _operationCompletedResult = new ContextAwareResult(_transport.IdentityRequired, true, null, this, s_contextSafeCompleteCallback);
                         lock (_operationCompletedResult.StartPostingAsyncOp())
                         {
-                            if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"Calling BeginConnect. Transport: {_transport}");
-                            _transport.BeginGetConnection(_operationCompletedResult, ConnectCallback, _operationCompletedResult, Host!, Port);
+                            if (_transport.IsConnected)
+                            {
+                                try
+                                {
+                                    SendMailAsync(_operationCompletedResult);
```

Comment:
It worries me that we will now have more/different code under the lock.

Reply:
It is what it is. The best way is to rewrite everything to Tasks, but with `SmtpClient` being deprecated, I can't think of anything better.